### PR TITLE
[ENH] Implementation of Extended Isolation Forest (EIF) anomaly detector

### LINKED ISSUE #2113: [ENH] Implement Extended Isolation Forest
### Describe the feature or idea you want to propose

The Extended Isolation Forest improves upon the Isolation Forest method.

### Describe your proposed solution

Implement directly in aeon:

- Paper: https://doi.org/10.1109/TKDE.2019.2947676
- Code: https://github.com/sahandha/eif
- Code (in TimeEval): https://github.com/TimeEval/TimeEval-algorithms/tree/main/eif

However, support semi-supervised as well as unsupervised usage similar to the [`PyODAdapter`](https://github.com/aeon-toolkit/aeon/blob/main/aeon/anomaly_detection/_pyodadapter.py) or [`KMeansAD`](https://github.com/aeon-toolkit/aeon/blob/main/aeon/anomaly_detection/_kmeans.py)

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Akhil-Jasson**
This PR implements the Extended Isolation Forest (EIF) algorithm.

#### Reference Issues/PRs
Fixes #2113

#### Does your contribution introduce a new dependency? If yes, which one?

Yes, it introduces H20.ai as a new dependency

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you after the PR has been merged.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [x] I've added the estimator/function to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [x] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

**Comment by MatthewMiddlehurst** [comment]
Please fill out the template and use the correct title format.

**Comment by Akhil-Jasson** [comment]
I've fixed the title format. Shall I proceed with adding EIF to the online API documentation?

**Comment by Ramana-Raja** [comment]
Hi @Akhil-Jasson, I checked out your PR, and the code looks great! That said, I don’t think using H2O is the best approach since Aeon doesn’t rely on it. It might be better to have our own implementation instead.

A few updates to consider:
   1. Could you update the section "Does your contribution introduce a new dependency?" and mention H2O there?

    2.The test cases seem to be missing—could you add them?

    3.Instead of importing the entire H2O module, it’s better to import only what’s needed to keep things lightweight.

**Comment by MatthewMiddlehurst** [comment]
New dependencies should be put in pyproject.toml otherwise this won't be tested, Still bits missing from the template

**Comment by Akhil-Jasson** [comment]
I've added the h2o dependency to pyproject.toml, but I'm encountering errors when running the test files. The test attempts to import aeon.anomaly_detection._eif but fails, indicating that the module doesn’t exist yet.  

Is there a step I'm missing for adding new modules to aeon? What could be the possible issue?

**Comment by MatthewMiddlehurst** [comment]
your import is incorrect.

**Review by SebastianSchmidl — CHANGES_REQUESTED** [review]
AI2O looks like a massive package. Do we actually want to include it as a dependency? The issue clearly states that we are looking for an implementation in aeon directly.

I know this is not mentioned in the corresponding issue, but I think it makes sense to work with sliding windows in EIF as well. We can always get the original behavior back by setting the window-size to 1.

File: aeon/anomaly_detection/_eif.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+    window_size : int, default=1
+        Size of the sliding window. If 1, the original point-wise behavior is used.
+    stride : int, default=1
+        Stride of the sliding window.
+    axis : int, default=1
+        The time point axis of the input series if it is 2D. If ``axis==0``, it is
+        assumed each column is a time series and each row is a time point. i.e. the
+        shape of the data is ``(n_timepoints, n_channels)``. ``axis==1`` indicates
+        the time series are in rows, i.e. the shape of the data is
+        ``(n_channels, n_timepoints)``.
```
Comment (SebastianSchmidl) [inline aeon/anomaly_detection/_eif.py:51]: Different data layouts are handled in the base class. Please do not expose an additional `axis`-parameter here.

File: aeon/anomaly_detection/_eif.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+    def _fit(self, X, y=None):
+        """Fit the model using X as training data.
+
+        Parameters
+        ----------
+        X : np.ndarray
+            Training data of shape (n_timepoints,) or (n_timepoints, n_channels)
+        y : np.ndarray, optional
+            Labels for semi-supervised learning. 0 for normal, 1 for anomalous.
+            If None, unsupervised learning is used.
```
Comment (SebastianSchmidl) [inline aeon/anomaly_detection/_eif.py:93]: Semi-supervised means training on normal data only, so just `0` is allowed here. Otherwise, this would be supervised training with normal data and known anomalies. Details: https://www.aeon-toolkit.org/en/stable/examples/anomaly_detection/anomaly_detection.html#Anomaly-Detection-in-aeon

File: aeon/anomaly_detection/_eif.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+
+        Returns
+        -------
+        self : object
+            Fitted estimator.
+        """
+        # Ensure X is 2D
+        if X.ndim == 1:
+            X = X.reshape(-1, 1)
```
Comment (SebastianSchmidl) [inline aeon/anomaly_detection/_eif.py:102]: This should not be necessary anymore if you remove the axis-parameter from the class and use a single data layout within the estimator.

File: aeon/anomaly_detection/_eif.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+
+            # Build tree
+            tree = self._build_tree(X_sample, rng)
+            self.forest_.append(tree)
+
+        # Calculate anomaly scores
+        scores = self._predict(X)
+
+        # Set threshold
+        self.threshold_ = float(np.percentile(scores, 100 * (1 - self.contamination)))
```
Comment (SebastianSchmidl) [inline aeon/anomaly_detection/_eif.py:156]: This looks suspicious. Why do you predict in `fit`?

File: aeon/anomaly_detection/_eif.py [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+        # Convert to anomaly scores
+        scores = 2 ** (-scores / self._c(self.sample_size_))
+
+        # If windowing was applied, we need to map scores back to original samples
+        if self.window_size > 1:
+            scores = reverse_windowing(
+                scores,
+                window_size=self.window_size,
+                stride=self.stride,
+                n_timepoints=self._n_samples_orig,
```
Comment (SebastianSchmidl) [inline aeon/anomaly_detection/_eif.py:300]: This just works if either EIF is used in an unsupervised way or both training and prediction time series have the same length!

File: aeon/anomaly_detection/_eif.py [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+        Returns
+        -------
+        float
+            Average path length
+        """
+        if n <= 1:
+            return 0
+        elif n == 2:
+            return 1
+        return 2 * (np.log(n - 1) + 0.5772156649) - 2 * (n - 1) / n
```
Comment (SebastianSchmidl) [inline aeon/anomaly_detection/_eif.py:259]: Where is the magic number from? Please at least add a comment.

**Review by SebastianSchmidl — CHANGES_REQUESTED** [review]
The code does not look too complicated, so I think, it was a good decision to not include H2O.ai as a dependency!

Thanks for the implementation effort. Please address the issues below and then compare your implementation to the original implementation by Hariri et al. to demonstrate that your/our implementation produces the same results as theirs.

I think this estimator is a suitable candidate for improving the performance with Numba / JITs. But let us first focus on making the implementation correct, and then tackle the performance optimization in another PR.

**Comment by Akhil-Jasson** [comment]
Previously, the tests were triggered automatically on commit based on the old repo structure. Since then, the structure has changed significantly. I ran the tests locally using the earlier setup, and they passed.

To accommodate the requirements of the test files—such as the need for y in the fit function—I’ve made it an optional parameter. I’ll take a look into the other comments as well.

Given the extent of changes in the repo structure, is it easier to just create a new branch from the updated main branch or is there an easier alternative? Please let me know!

**Comment by MatthewMiddlehurst** [comment]
I would merge main into your branch, the bulk of your code should not conflict. It is mainly that we have refactored the module so it has all moved around a bit. If you do not know which module to put your implementation in just put it anywhere for now and we can discuss after where is best.

**Comment by Akhil-Jasson** [comment]
Changes have been made to this branch to align with the newly refactored module structure. I'll soon address the feedback to remove the axis hyperparameter and add the necessary comments as requested.

**Comment by MatthewMiddlehurst** [comment]
Hello, is this still active?

**Comment by Akhil-Jasson** [comment]
Hello @MatthewMiddlehurst,
Thank you for checking in. I apologize for the long pause. I've had some time away, but I’m still eager to continue work on the Extended Isolation Forest implementation.

**Comment by Akhil-Jasson** [comment]
Hi @SebastianSchmidl  and @MatthewMiddlehurst,

I've addressed the feedback from the previous review:

Key Changes Made:
- Removed threshold calculation from _fit()
- Fixed sliding window length handling - Now works with different-sized prediction data (addresses the length dependency issue)
- Added comments for magic numbers - Explained the Euler-Mascheroni constant in _c() method
- Maintained unsupervised/semi-supervised support - Works with both `y=None` and `y=0` modes

Please let me know if there are any remaining concerns.

Thanks for the detailed feedback, it really helped improve the implementation!

**Comment by MatthewMiddlehurst** [comment]
Hello, based on the current state of the PR and the changes made, I think this is going to take a lot more effort to get anywhere. If you want to continue, please:

1. Look at the series anomaly detection base class. There is a misunderstanding here of what we mean by unsupervised and semi-supervised and how that is configured in the framework. The notebook we linked appears to contradict this, so that is our bad.
2. Look at other anomaly detections to see the format and tags used. The init requires you to provide an axis to the base class which determines the shape of the data
3. Merge to the most recent main branch so testing can be run and the code is up to date with recent changes. There is also no testing for this currently.
4. Run your implementation and compare against the original results

This is a good base, but we cant accept code which breaks the framework or we are unsure works as intended.

File: aeon/anomaly_detection/outlier_detection/_eif.py [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+                scores = scores[:original_n_timepoints]
+            elif len(scores) < original_n_timepoints:
+                # Pad with edge values if needed
+                pad_length = original_n_timepoints - len(scores)
+                scores = np.pad(scores, (0, pad_length), mode="edge")
+
+        return scores
+
+    def predict_labels(self, X) -> np.ndarray:
+        """Predict if points are anomalies or not.
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/outlier_detection/_eif.py:322]: this method shouldnt exist IMO.

File: aeon/anomaly_detection/outlier_detection/_eif.py [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+class EIF(BaseAnomalyDetector):
+    """Extended Isolation Forest (EIF) for anomaly detection.
+
+    Implementation of the Extended Isolation Forest algorithm that uses
+    random hyperplanes for splitting instead of axis-parallel splits.
+    This allows for better handling of high-dimensional data and complex distributions.
+
+    The implementation supports both unsupervised and semi-supervised learning:
+    - Unsupervised: No labels provided (y=None)
+    - Semi-supervised: Labels provided (y=0 for normal data)
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/outlier_detection/_eif.py:21]: by unsupervised we mean it makes predictions on the same data with no labels, for semi-supervised it fits on data then this can be used for other data. I do not think this is unsupervised in our framework.

File: aeon/anomaly_detection/outlier_detection/_eif.py [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+        -------
+        self : object
+            Fitted estimator.
+        """
+        if y is not None:
+            y = np.asarray(y)
+            if not np.all(y == 0):
+                raise ValueError(
+                    "Semi-supervised mode only supports training on normal data (y==0)."
+                )
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/outlier_detection/_eif.py:98]: I do not think this is needed as y is never used. This may be worth putting in the base class, however.

File: aeon/anomaly_detection/outlier_detection/_eif.py [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+            X_windowed = X
+
+        # Set random state
+        rng = np.random.RandomState(self.random_state)
+
+        # Determine sample size
+        n_samples = X_windowed.shape[0]
+        if isinstance(self.sample_size, str):
+            if self.sample_size != "auto":
+                raise ValueError("sample_size must be 'auto', int or float")
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/outlier_detection/_eif.py:115]: this error catching could be better. it will only run if it is an incorrect str

File: aeon/anomaly_detection/outlier_detection/_eif.py [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+        # Apply sliding window if window_size > 1
+        if self.window_size > 1:
+            X_windowed, _ = sliding_windows(
+                X, window_size=self.window_size, stride=self.stride
+            )
+        else:
+            X_windowed = X
+
+        # Set random state
+        rng = np.random.RandomState(self.random_state)
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/outlier_detection/_eif.py:109]: can you use https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_random_state.html for rng, see the docstring of other algorithms which use rng for a better one

File: aeon/anomaly_detection/outlier_detection/_eif.py (hunk 8 above)
Comment (MatthewMiddlehurst) [reply aeon/anomaly_detection/outlier_detection/_eif.py:21]: We made a decision previously that different types i.e. unsupervised and semi-supervised should be different classes. I would just go for semi-supervised for now.

**Comment by TonyBagnall** [comment]
hi @Akhil-Jasson in the spirit of housekeeping I am converting this to draft, feel free to reopen if you want to continue with it

**Comment by TonyBagnall** [comment]
made redundant by #3626
